test/e2e/addoninstall_test.go

```diff
@@ -0,0 +1,149 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"open-cluster-management.io/addon-framework/pkg/addonmanager/constants"
+	addonapiv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+var _ = ginkgo.Describe("Testing installation of work-manager add-on", func() {
+	var clusterName string
+	var annotations map[string]string
+	addonName := "work-manager"
+	hostingClusterName := "local-cluster"
+
+	ginkgo.JustBeforeEach(func() {
+		cluster, err := clusterClient.ClusterV1().ManagedClusters().Get(context.Background(), clusterName, metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			cluster = &clusterv1.ManagedCluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        clusterName,
+					Annotations: annotations,
+				},
+				Spec: clusterv1.ManagedClusterSpec{
+					HubAcceptsClient: true,
+				},
+			}
+			_, err = clusterClient.ClusterV1().ManagedClusters().Create(context.Background(), cluster, metav1.CreateOptions{})
+			gomega.Expect(err).ToNot(gomega.HaveOccurred())
+			return
+		}
+
+		gomega.Expect(err).ToNot(gomega.HaveOccurred())
+		if reflect.DeepEqual(cluster.Annotations, annotations) {
+			return
+		}
+
+		cluster = cluster.DeepCopy()
+		cluster.Annotations = annotations
+		_, err = clusterClient.ClusterV1().ManagedClusters().Update(context.Background(), cluster, metav1.UpdateOptions{})
+		gomega.Expect(err).ToNot(gomega.HaveOccurred())
+	})
+
+	ginkgo.JustAfterEach(func() {
+		err := clusterClient.ClusterV1().ManagedClusters().Delete(context.Background(), clusterName, metav1.DeleteOptions{})
+		gomega.Expect(err).ToNot(gomega.HaveOccurred())
+	})
+
+	haveNoAddonIt := func() {
+		ginkgo.It("should have no add-on installed", ginkgo.Offset(1), func() {
+			gomega.Consistently(func() bool {
+				_, err := addonClient.AddonV1alpha1().ManagedClusterAddOns(clusterName).Get(context.Background(), addonName, metav1.GetOptions{})
+				return errors.IsNotFound(err)
+			}, 30, 2).Should(gomega.BeTrue())
+		})
+	}
+
+	haveAddonInDefaultModeIt := func() {
+		ginkgo.It("should have add-on installed in default mode", ginkgo.Offset(1), func() {
+			var addon *addonapiv1alpha1.ManagedClusterAddOn
+			var err error
+			gomega.Eventually(func() error {
+				addon, err = addonClient.AddonV1alpha1().ManagedClusterAddOns(clusterName).Get(context.Background(), addonName, metav1.GetOptions{})
+				return err
+			}, eventuallyTimeout, eventuallyInterval).ShouldNot(gomega.HaveOccurred())
+
+			gomega.Expect(addon).ShouldNot(gomega.BeNil())
+			gomega.Expect(addon.Annotations).ShouldNot(gomega.HaveKey(addoninstall.AnnotationKlusterletHostingClusterName))
+			gomega.Expect(addon.Spec.InstallNamespace).To(gomega.Equal(addoninstall.DefaultAddOnInstallNamespace))
+		})
+	}
+
+	ginkgo.Context("cluster is imported in default mode", func() {
+		ginkgo.When("add-on installation is disabled", func() {
```

Comment:
Why do we use a `ginkgo.When` here? it would make the hierarchy 4 levels(Describe-context-when-it), but we recommend 3 levels, right？

Reply:
In a previous discussion, I proposed to group test cases with `Describe` or `Context`. And  we can treat `Describe` as feature and `Context` as sub-feature. Actually the existing test cases are not been organized in this way. And sometimes a test case may cover several features, which make it hard to organize test case in this way. 

Since we upgraded to Ginkgo v2, we might be able to leverage a new feature `label` decorator to help us associate features with test cases. It can maintain a m:n relationship between features (labels) and test cases. For example,

```golang
Describe("Testing addon Installation", Label("add-on", "add-on install"), func() {
  It("should install in hosted mode", Label("hosted mode"), func() {
    // has labels [add-on, add-on install, hosted mode]
  })
})
```

Let's discuss this in a separated issue or a meeting.